// find-nearest-package-json/src/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.findNearestPackageJson = undefined;

let findNearestPackageJson = exports.findNearestPackageJson = (() => {
  var _ref = _asyncToGenerator(function* (directoryPath = path.resolve()) {
    try {
      const packageJsonPath = path.join(directoryPath, 'package.json');
      const packageJsonData = JSON.parse((yield readFile(packageJsonPath)));
      return {
        path: packageJsonPath,
        data: packageJsonData
      };
    } catch (error) {
      const parentDirectoryPath = path.dirname(directoryPath);
      if (parentDirectoryPath === directoryPath) {
        throw new Error('No package.json files found');
      }
      return findNearestPackageJson(parentDirectoryPath);
    }
  });

  return function findNearestPackageJson() {
    return _ref.apply(this, arguments);
  };
})();

exports.findNearestPackageJsonSync = findNearestPackageJsonSync;

var _fs = require('fs');

var fs = _interopRequireWildcard(_fs);

var _path = require('path');

var path = _interopRequireWildcard(_path);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

function readFile(filePath) {
  return new Promise((resolve, reject) => {
    fs.readFile(filePath, 'utf8', (error, result) => {
      if (error) reject(error);else resolve(result);
    });
  });
}

function findNearestPackageJsonSync(directoryPath = path.resolve()) {
  try {
    const packageJsonPath = path.join(directoryPath, 'package.json');
    const packageJsonData = JSON.parse(readFileSync(packageJsonPath));
    return {
      path: packageJsonPath,
      data: packageJsonData
    };
  } catch (error) {
    const parentDirectoryPath = path.dirname(directoryPath);
    if (parentDirectoryPath === directoryPath) {
      throw new Error('No package.json files found');
    }
    return findNearestPackageJsonSync(parentDirectoryPath);
  }
}

function readFileSync(filePath) {
  return fs.readFileSync(filePath, 'utf8');
}
